src/lib/arch/riscv/Instruction_decode.cc

```diff
@@ -0,0 +1,254 @@
+#include "InstructionMetadata.hh"
+#include "simeng/arch/riscv/Architecture.hh"
+#include "simeng/arch/riscv/Instruction.hh"
+
+namespace simeng {
+namespace arch {
+namespace riscv {
+
+/********************
+ * HELPER FUNCTIONS
+ *******************/
+
+/** Parses the Capstone `riscv_reg` value to generate an architectural register
+ * representation.
+ *
+ * WARNING: this conversion is FRAGILE, and relies on the structure of the
+ * `riscv_reg` enum. Updates to the Capstone library version may cause this to
+ * break. */
+Register csRegToRegister(unsigned int reg) {
+  // Check from top of the range downwards
+
+  if (RISCV_REG_X31 >= reg && reg >= RISCV_REG_X1) {
+    // Capstone produces 1 indexed register operands
+    return {RegisterType::GENERAL, static_cast<uint16_t>(reg - 1)};
+  }
+
+  if (reg == RISCV_REG_X0) {
+    // Zero register
+    return Instruction::ZERO_REGISTER;
+  }
+
+  assert(false && "Decoding failed due to unknown register identifier");
+  return {std::numeric_limits<uint8_t>::max(),
+          std::numeric_limits<uint16_t>::max()};
+}
+
+// Invalidate instructions that are currently not yet implemented. This prevents
+// errors during speculated branches with unknown destinations; non-executable
+// memory is decoded into valid but not implemented instructions tripping
+// assertions.
+// TODO remove once all extensions are supported
+void Instruction::invalidateIfNotImplemented() {
+  if (metadata.opcode >= Opcode::RISCV_ADD &&
+      metadata.opcode <= Opcode::RISCV_BNE)
+    return;
+  if (metadata.opcode >= Opcode::RISCV_DIV &&
+      metadata.opcode <= Opcode::RISCV_ECALL)
+    return;
+  if (metadata.opcode >= Opcode::RISCV_JAL &&
+      metadata.opcode <= Opcode::RISCV_SD)
+    return;
+  if (metadata.opcode >= Opcode::RISCV_SH &&
+      metadata.opcode <= Opcode::RISCV_SRAW)
+    return;
+  if (metadata.opcode >= Opcode::RISCV_SRL &&
+      metadata.opcode <= Opcode::RISCV_SW)
+    return;
+  if (metadata.opcode >= Opcode::RISCV_XOR &&
+      metadata.opcode <= Opcode::RISCV_XORI)
+    return;
+  if (metadata.opcode == Opcode::RISCV_FENCE) return;
+
+  exception_ = InstructionException::EncodingUnallocated;
+  exceptionEncountered_ = true;
+  return;
+}
+
+/******************
+ * DECODING LOGIC
+ *****************/
+void Instruction::decode() {
+  invalidateIfNotImplemented();
+  if (exceptionEncountered_) return;
+  if (metadata.id == RISCV_INS_INVALID) {
```

Comment:
invalidateIfNotImplemented is a function used as a work around for speculated instructions fetched from non-executable memory whereas the RISCV_INS_INVALID checks whether the encoding is decoded as such. It is the intention that invalidateIfNotImplemented is eventually removed but the INVALID check will remain. 

While they provide the same functionality the reasoning for them is different and so is the future of their existence, which is why they were kept separate. Will keep as is to avoid accidentally removing at a later date, unless you feel strongly otherwise.